### PR TITLE
[kvutils] Record all key lookups by transaction committer [KVL-750]

### PR DESCRIPTION
**Pull request by miklos-da**
*Motivation*: we didn't record contract key lookups previously hence the read-set produced for pre-execution was not correct.

Summary of changes:
* Added `CommitContext.read` to read from the input state without throwing if the key has not been declared as a submission input.
* Route all state read calls via the previous method in `TransactionCommitter`.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by miklos-da** [comment]
FYI @dajmaki @meiersi-da @mziolekda.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:237]: I'd prefer either testing the property at this level (i.e. testing `validateModelConformance` for it rather than `CommitContext.read`) or encapsulating `inputs` in some way, so that `TransactionCommitter` changes cannot cause a regression either.

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:237]: I agree with hiding and encapsulating `inputs`. It's clear that direct access is dangerous.

**Review by andreaslochbihler-da — COMMENTED** [review]
The previous behaviour was indeed wrong. A submitter could have omitted a contract key from a negative `LookupByKey` node and we wouldn't have noticed. 

I'm a bit worried about the "not throwing" aspect if an input state key is missing. For if the declared input keys are used to pre-fetch the inputs from a DB, then we won't have fetched the missing input state key. So how does conflict detection on the underlying storage system know what version to assume for keys in the read set that have never been fetched from the DB? Is the assumption that those keys must be absent in the KV store?

**Comment by miklos-da** [comment]
> The previous behaviour was indeed wrong. A submitter could have omitted a contract key from a negative `LookupByKey` node and we wouldn't have noticed.
> 
> I'm a bit worried about the "not throwing" aspect if an input state key is missing. For if the declared input keys are used to pre-fetch the inputs from a DB, then we won't have fetched the missing input state key. So how does conflict detection on the underlying storage system know what version to assume for keys in the read set that have never been fetched from the DB? Is the assumption that those keys must be absent in the KV store?

The underlying storage system can rely on the fingerprints it provides us. We handle the corner cases as follows:
* If a key is declared as an input key but is not available in the DB, we pass back the fingerprint we got for it as-is in the read-set.
* If a key is not declared as an input and the committer tries to read it, we throw an exception because we don't have a fingerprint for it to return in the read-set.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:237]: That's a great idea -- I'll encapsulate `inputs` so that we can be certain we record all read keys.

**Comment by andreaslochbihler-da** [comment]
> The underlying storage system can rely on the fingerprints it provides us for conflict detection. We handle the corner cases as follows:
> 
>     * If a key is declared as an input key but is not available in the DB, we pass back the fingerprint we got for it as-is in the read-set.
> 
>     * If a key is not declared as an input and the committer tries to read it, we throw an exception because we don't have a fingerprint for it to return in the read-set (see [PreExecutingSubmissionValidator.generateReadSet](https://github.com/digital-asset/daml/blob/2b67c74d9a29e024ab7f3263b73afb11a21f7c5d/ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/validator/preexecution/PreExecutingSubmissionValidator.scala#L134)).

Sounds good.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala
Comment (gerolf-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: Should we consider adding a `collectFiltered` (not a good name, I know) to only do a single pass? It also looks like we're recording a larger set (all non empty contractIds) than what we actually need (only those that have a contract key).

Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: Yes, that should work as well. When do contract IDs not have an associated contract key? I.e., in which scenarios would cause this a difference in the returned read-set?

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala:72]: Minor: an utility here could improve readability (keys are built thrice only in this test).

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala:80]: Minor: naming the parameter does not improve readability here.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala
Comment (SamirTalwar-DA) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:23]: It's really weird that this is a trait. Perhaps it should be a class and accept `inputs`, `recordTime`, and `participantId` as parameters?

Definitely something for a subsequent changeset, though.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: I think it's a good idea to have a function that reads many inputs using a partial function. Perhaps `collectInputs` would be a decent name, as it's effectively a wrapped version of `.inputs.collect { ... }` with a side effect.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: If there is a need to filter on values, another option would be to allow filtering on them as well or to provide use-case specific encapsulations of `inputs`.

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: BTW, I am fairly sure most contracts don't have keys; you have to add a contract key explicitly with the DAML `key` keyword.

**Review by gerolf-da — APPROVED** [review]

**Review by fabiotudone-da — APPROVED** [review]
LGTM modulo @gerolf-da's [comment here](https://github.com/digital-asset/daml/pull/8082/files#r533530562): once that is solved it's good to go from my perspective. Thanks for encapsulating `inputs`!

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala
Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: @fabiotudone-da: If you filter out a value, do you still need to mark it as read?

If so, I'd prefer if that was somehow separated in the API.

Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: sorry, I edited the original comment because github doesn't auto update without refreshing.
Yes, if you filter on the value, you must have read the value to be able to do that.

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: > @fabiotudone-da: If you filter out a value, do you still need to mark it as read?

No, AFAIU.

Comment (gerolf-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: filter ON != filter OUT

Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: @SamirTalwar Ignore my comment, @gerolf-da is right, because the filtering is logical.

Comment (SamirTalwar-DA) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:216]: Sorry, that's what I meant. Filtering out _based on_ values.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContext.scala:23]: Yes, it is -- we can fix that in a later PR.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala]: Replaced by `collectInputs` that takes a partial function to do the filtering/transformation.

File: ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala
Comment (miklos-da) [reply ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/committer/CommitContextSpec.scala]: You have to name it as it is the second optional parameter.

**Pull request merged**
